File: 프로그래머스/Level3/네트워크.py
# https://programmers.co.kr/learn/courses/30/lessons/43162

from collections import deque

def solution(n, computers):
    visited = [False] * n
    answer = 0
    for i in range(n):
        if visited[i] == False:
            answer += 1
            q = deque([i])
            visited[i] = True
            while q:
                temp = q.popleft()
                info = computers[temp]
                for j in range(len(info)):
                    if i != j and visited[j] == False and info[j] == 1:
                        visited[j] = True
                        q.append(j)
    return answer

File: 프로그래머스/Level3/네트워크.js
// https://programmers.co.kr/learn/courses/30/lessons/43162

const solution = (n, computers) => {
  const visited = new Array(n).fill(false);
  let cnt = 0;
  for (let i = 0; i < n; i++) {
    if (visited[i]) continue;
    cnt++;
    const q = [i];
    while (q.length !== 0) {
      const next = q.shift();
      visited[next] = true;
      const adj = computers[next];
      adj.forEach((isConnected, idx) => {
        if (isConnected === 1 && !visited[idx]) q.push(idx);
      });
    }
  }
  return cnt;
};
